### ISSUE TITLE
Integrate CuDNN in Theano

### ISSUE THREAD

**Issue by nouiz** [Theano/Theano#2094]
https://developer.nvidia.com/cuDNN
- [x] Test with R2 RC2 and RC3
- [x] integrate the convolution/cross-correlation kernel
- [x] Softmax forward
- [x] Softmax backward
- [x] integrate the pooling (Maximum and Average)
- [x] check if we can implement ignore_border=False.
  - Need explicit padding in the graph
- [x] Add infer_shape methods (close to be done)
- [x] remove gpu_contiguous (partly done in gh-2551, but also need gh-2421)
- [x] For the convolution, make the pad argument available to the user.
  - [x] allor border_mode=half/same
- [x] dnn avail detection should use NVCC_compiler. GCC do not work on Windows and we worked around on Linux by manually adding the paths.
- [x] add support for new CuDNN R2 features.
  - [x] Add support for the padding parameter.
  - [x] allow alpha/beta usage
  - [x] opt to do inplace update
  - [x] Allow faster forward version with extra memory.
  - [x] conv 3d
- [x] Compile less module. Do not hardcode the stirdes/filter size in the c code.
- [ ] Always our cudnnGetErrorString version. In version where it exist in cudnn, if we do not know the error code, call cudnn version. Our version give better error message in some cases.
- [x] BatchNormalization (cudnn v4) (speed up?)
- [ ] LRN (speed up?)
- [ ] LCN (speed up?)

CPU interface
- [x] CPU conv/corr interface (gh-2665)
- [ ] CPU softmax interface (different softmax layout, combination)
- [x] CPU avg pooling

Probably not needed:
- [ ] Tensor transformation functions (if needed? Is it optimized copy?)
- [ ] Neuron activations forward and backward (Rectified Linear, Sigmoid, Hyperbolic Tangent)

**Comment by f0k** [comment]
I've just seen the following among the "Known Issues" of the RC2 release notes:
> Several cudnn*Backward() routines assume that the corresponding cudnn*Forward() routine
will have been called previously without documenting this assumption.

I'll ask which routines exactly this applies to, and what happens if you break this assumption. I know that we may at least use some of the convolution backward routines before using the corresponding forward routine, but I hope the tests would have caught this if there was any problem. Anyway, we should keep an eye on that.

**Comment by cancan101** [comment]
Is there something along the lines of `theano.sandbox.cuda.dnn.dnn_available()` that can be used to tell what version of cuDNN is being used? I would love to add to my logs a printout of the version of cuDNN in use for debugging purposes.

**Comment by nouiz** [comment]
...dnn.version()

Fred

**Comment by themummy** [comment]
Did anybody have any luck using cuDNN on Windows? I'm stuck at getting a DLL Import error.

**Comment by f0k** [comment]
> Did anybody have any luck using cuDNN on Windows?

You may want to check on [theano-users](https://groups.google.com/forum/#!forum/theano-users) or re-post your question there. github issues are meant for bug reports, feature proposals and submitting patches. If it turns out to be a bug, please file a new Issue, this one mostly hosts a Todo list. Cheers!

**Comment by themummy** [comment]
Thanks, @f0k!

**Comment by nouiz** [comment]
We did all what we need for 0.9, I remove the milestone.
